### PR TITLE
[SpeechLM] Parallel strategies.

### PR DESCRIPTION
**Pull request by jctian98**
# Add parallel_utils for SpeechLM tensor/pipeline parallelism

## Summary

Adds a new `parallel_utils` module under `espnet2/speechlm/model/speechlm/` that provides model-specific parallelization strategies for SpeechLM training.

- **`init_parallel_dims`** — initializes parallel dimensions (TP/PP/DP) from the titan config
- **`build_pipeline`** — constructs pipeline-parallel schedules for SpeechLM
- **`parallelize_qwen3_hf`** — FSDP2 wrapping tailored to the Qwen3 HF architecture
- **`GroupedMoE`** — grouped Mixture-of-Experts layer
- **`parallel_strategies`** registry — `{"qwen3": parallelize_qwen3_hf, ...}` so new model series can plug in a `(model, parallel_dims, titan_config) -> model` function

## Files added

| File | Purpose |
| --- | --- |
| `espnet2/speechlm/model/speechlm/parallel_utils/__init__.py` | Public API and strategy registry |
| `espnet2/speechlm/model/speechlm/parallel_utils/parallel_dims.py` | Parallel dimension initialization |
| `espnet2/speechlm/model/speechlm/parallel_utils/pipeline.py` | Pipeline parallel builder |
| `espnet2/speechlm/model/speechlm/parallel_utils/qwen3.py` | Qwen3 FSDP2 parallelization |
| `espnet2/speechlm/model/speechlm/parallel_utils/grouped_moe.py` | Grouped MoE implementation |

Total: **5 files, +1223 lines**.

## Usage

```python
from espnet2.speechlm.model.speechlm.parallel_utils import (
    init_parallel_dims,
    parallel_strategies,
)

parallel_dims, local_rank, global_rank = init_parallel_dims(titan_config)
parallelize_fn = parallel_strategies["qwen3"]
model = parallelize_fn(model, parallel_dims, titan_config)

### CODE REVIEW

File: espnet2/speechlm/model/speechlm/parallel_utils/qwen3.py [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+    Returns:
+        Model with MoE blocks replaced
+    """
+    ep_enabled = parallel_dims is not None and parallel_dims.ep_enabled
+
+    has_moe = False
+    for idx, layer in enumerate(model.model.layers):
+        if isinstance(layer, nn.Identity):
+            continue
+        layer = layer.cuda()
```
[the remark this answers is not in the document]
Comment (jctian98) [reply espnet2/speechlm/model/speechlm/parallel_utils/qwen3.py:182]: this is intentional as we want to move and shard the model progressively rather than in one go.

File: espnet2/speechlm/model/speechlm/parallel_utils/grouped_moe.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+        # Workaround: EP>1 backward produces aten.index_add (via index_select
+        # backward in DeepEP's _permute_tokens). Inductor has a conditional
+        # decomp for index_add that returns NotImplemented for bf16, but its
+        # mere presence in the decomp table conflicts with make_fallback on
+        # newer PyTorch (e.g. required for Blackwell/sm_100). Removing it lets
+        # Inductor use the ATen kernel directly as a fallback — same behavior
+        # the bf16 decomp was already falling back to anyway.
+        import torch._inductor.decomposition as _ind_decomp
+
+        _ind_decomp.decompositions.pop(torch.ops.aten.index_add.default, None)
```
[the remark this answers is not in the document]
Comment (jctian98) [reply espnet2/speechlm/model/speechlm/parallel_utils/grouped_moe.py:348]: this is intentional to be compatible with moe dynamics during training

File: espnet2/speechlm/model/speechlm/parallel_utils/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+    # Override _split_inputs to pass through pre-split microbatch lists.
+    # The trainer passes (arg_mbs, kwarg_mbs) as the two positional args
+    # to step(). Inside step(), these become args=(arg_mbs, kwarg_mbs).
+    # _split_inputs must return (args_split, kwargs_split) where
+    # args_split is the list of per-microbatch arg tuples and
+    # kwargs_split is the list of per-microbatch kwarg dicts.
+    # Works identically for both PipelineScheduleSingle and
+    # PipelineScheduleMulti since both call _split_inputs the same way.
+    schedule._split_inputs = lambda args, kwargs: (args[0], args[1])
```
[the remark this answers is not in the document]
Comment (jctian98) [reply espnet2/speechlm/model/speechlm/parallel_utils/pipeline.py:146]: this is to allow the dict-based input rather than the tensor based method. multimodal is not well supported by current torch

File: espnet2/speechlm/model/speechlm/parallel_utils/qwen3.py [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+        When EP is enabled and the layer is MoE, wraps the
+        already-sharded experts with FSDP on edp_mesh before wrapping
+        the full layer on dp_mesh.
+        """
+        layer.to(device)
+
+        if ep_enabled and getattr(layer.mlp, "ep_enabled", False):
+            efsdp_size = edp_mesh["efsdp"].size()
+            assert efsdp_size == 1, "FSDP on experts is not supported yet."
+            fully_shard(layer.mlp.experts, **fsdp_ep_config)
```
[the remark this answers is not in the document]
Comment (jctian98) [reply espnet2/speechlm/model/speechlm/parallel_utils/qwen3.py:295]: this keep it for now. it's enough for current model scale.

File: espnet2/speechlm/model/speechlm/parallel_utils/qwen3.py [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+        vpp_index: Virtual pipeline stage index on this rank. Used as
+            fallback when ``model.stage_idx`` is not set. When
+            ``ac_config`` is a list, it is indexed by
+            ``model.stage_idx`` (the global virtual stage index).
+
+    Returns:
+        Model with activation checkpointing applied
+    """
+    if isinstance(ac_config, list):
+        stage_idx = getattr(model, "stage_idx")
```
[the remark this answers is not in the document]
Comment (jctian98) [reply espnet2/speechlm/model/speechlm/parallel_utils/qwen3.py:383]: cannot use vpp_index for stage_idx. they are different

**Comment by jctian98** [comment]
@sw005320 @wanchichen @Masao-Someki 

The next PR

**Comment by wanchichen** [comment]
The design looks good to me, don't have any concerns on the efficiency or other technical aspects

File: espnet2/speechlm/model/speechlm/parallel_utils/grouped_moe.py
Comment (sw005320) [inline espnet2/speechlm/model/speechlm/parallel_utils/grouped_moe.py:1]: This is a highly optimized code, and I learned a lot of techniques by reading this code :)

File: espnet2/speechlm/model/speechlm/parallel_utils/grouped_moe.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+    Qwen3MoeSparseMoeBlock,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def _route(gate, hidden_states_flat, top_k, norm_topk_prob):
+    """Top-k softmax routing shared by GroupedMoeBlock and GroupedMoeBlockEP."""
+    router_logits = gate(hidden_states_flat)
+    routing_weights = F.softmax(router_logits, dim=1, dtype=torch.float)
```
Comment (sw005320) [inline espnet2/speechlm/model/speechlm/parallel_utils/grouped_moe.py:51]: why `dtype=torch.float`?
(I thought that it's better to use bfloat or double since softmax tends to require more precision)

**Review by sw005320 — APPROVED** [review]
LGTM

Is this MoE actually used in Bagpiper or some future preparations?
Can you clarify it and add the comment to the appropriate place?

File: espnet2/speechlm/model/speechlm/parallel_utils/parallel_dims.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    ep = titan_config.get("ep", 1)
+
+    parallel_dims = ParallelDims(
+        dp_replicate=titan_config.get("dp_replicate", 1),
+        dp_shard=titan_config.get("dp_shard", -1),  # -1 = auto
+        cp=1,
+        tp=1,
+        pp=titan_config.get("pp_degree", 1),
+        ep=ep,
+        etp=1,
```
Comment (Masao-Someki) [inline espnet2/speechlm/model/speechlm/parallel_utils/parallel_dims.py:70]: I am not sure if it is fine to use this, as it seems this argument is duplicated in the latest torchtitan.
https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/parallel_dims.py

**Review by Masao-Someki — COMMENTED** [review]
I think this PR basically does not impact on espnet3, as it is basically utilities to get the model and output parallel model.
I found one concern, but other part are well documented/tested, they looks good to me.

**Comment by jctian98** [comment]
> LGTM
> 
> Is this MoE actually used in Bagpiper or some future preparations? Can you clarify it and add the comment to the appropriate place?

The code is currently not used in bagpiper as that was a Dense model, but it's in current training right now.

File: espnet2/speechlm/model/speechlm/parallel_utils/grouped_moe.py (hunk 6 above)
Comment (jctian98) [reply espnet2/speechlm/model/speechlm/parallel_utils/grouped_moe.py:51]: casting router logits to fp32 is a shared practice in deepspeed MoE and official HF implementation of Qwen3-MoE. It seems this practice is empirically important to the training stability of MoE.

https://github.com/huggingface/transformers/blob/b6fb4ec0f24cfad232de46a02441b9cc6362cd22/src/transformers/models/qwen3_moe/modeling_qwen3_moe.py#L266

File: espnet2/speechlm/model/speechlm/parallel_utils/parallel_dims.py (hunk 7 above)
Comment (jctian98) [reply espnet2/speechlm/model/speechlm/parallel_utils/parallel_dims.py:70]: Sorry, I'm a bit confused about this concern.
We are directly calling `ParallelDims` from torchtitan. Which one do you think duplicated?

Comment (Masao-Someki) [reply espnet2/speechlm/model/speechlm/parallel_utils/parallel_dims.py:70]: Oh I mean, it will be duplicated in the future version = cause error here once they have version update. That was my concern

**Comment by sw005320** [comment]
> > LGTM
> > Is this MoE actually used in Bagpiper or some future preparations? Can you clarify it and add the comment to the appropriate place?
> 
> The code is currently not used in bagpiper as that was a Dense model, but it's in current training right now.

Hmm.
In the future, I'd recommend you not include such experimental code.
This requires an additional review step (which may also be changed in the future if it is ongoing).

File: espnet2/speechlm/model/speechlm/parallel_utils/parallel_dims.py (hunk 7 above)
Comment (jctian98) [reply espnet2/speechlm/model/speechlm/parallel_utils/parallel_dims.py:70]: I see. For titan version that i'm using, the ETP is required with no default, so maybe we can keep it for now.
I see the newer version of titan removes this etp option; we can make a small patch to this in the future when titan gets more stable.

**Comment by jctian98** [comment]
@sw005320 CI fixed, maybe merge?
